cosmos/relayer
# CosmosChainProcessor query loop

Adds main query loop for `CosmosChainProcessor`

`Run` function will:
1. Initially, before starting loop, query for latest height, retrying until is successful. 

I went back and forth on whether this should return an error which would stop the `CosmosChainProcessor` or not, since once in the loop, we want it to continue to retry in the case of a node outage incident. But maybe initially it should panic?

2. Determine initial block to query from provided `initialBlockHistory`
3. Start the query loop

Each iteration of the query loop:
1. Query latest height
2. If not yet in sync, check if `latestHeight` - `latestQueriedBlock` is within `inSyncThreshold`. If determined to now be in sync, update concurrently accessed `inSync` property to true.
3. Query `BlockResults` from each block from `latestQueriedBlock + 1` to `latestHeight`. This loop defined as `BlockQuery` below.
4. If this is the first query loop iteration where `inSync == true`, trigger the associated `PathProcessors` to check if both chains are `InSync()`, and start processing the backlog of IBC messages if so.
5. Check how long this cycle of the query loop took. If less than `minQueryLoopDuration`, wait until `minQueryLoopDuration` has passed from the start of the cycle before continuing. TODO in a future PR, make `minQueryLoopDuration` dynamic based on block production time.

For each `BlockQuery` iteration:
1. Iterate through transactions in block.
2. Use event parser code from previous PR to pull out parsed IBC messages from successful transactions.
3. TODO handle IBC messages in future PR
